src/Interpreters/TLSLog.cpp

```diff
@@ -0,0 +1,218 @@
+#include <Interpreters/TLSLog.h>
+
+#include <Columns/ColumnArray.h>
+#include <Common/DateLUT.h>
+#include <Common/ThreadPool.h>
+#include <Core/NamesAndTypes.h>
+#include <DataTypes/DataTypeArray.h>
+#include <DataTypes/DataTypeDate.h>
+#include <DataTypes/DataTypeDateTime.h>
+#include <DataTypes/DataTypeDateTime64.h>
+#include <DataTypes/DataTypeEnum.h>
+#include <DataTypes/DataTypeLowCardinality.h>
+#include <DataTypes/DataTypeString.h>
+#include <DataTypes/DataTypesNumber.h>
+#include <Interpreters/SystemLog.h>
+#include <Poco/Delegate.h>
+#if USE_SSL
+#include <Poco/Net/SSLManager.h>
+#include <Access/Common/SSLCertificateSubjects.h>
+#include <Poco/Net/VerificationErrorArgs.h>
+#endif
+
+namespace
+{
+using namespace DB;
+
+auto eventTime()
+{
+    const auto finish_time = std::chrono::system_clock::now();
+
+    return std::make_pair(timeInSeconds(finish_time), timeInMicroseconds(finish_time));
+}
+
+void fillColumnArray(const Strings & data, IColumn & column)
+{
+    auto & array = typeid_cast<ColumnArray &>(column);
+    size_t size = 0;
+    auto & data_col = array.getData();
+    for (const auto & name : data)
+    {
+        data_col.insertData(name.data(), name.size());
+        ++size;
+    }
+    auto & offsets = array.getOffsets();
+    offsets.push_back(offsets.back() + size);
+}
+
+}
+
+namespace DB
+{
+
+TLSLogElement::TLSLogElement(const String &failure_reason_, uint64_t failure_count_)
+    : type(failure_reason_.empty() ? TLSLogElementType::SESSION_SUCCESS : TLSLogElementType::SESSION_FAILURE), failure_reason(failure_reason_), failure_count(failure_count_)
+{
+    std::tie(event_time, event_time_microseconds) = eventTime();
+}
+
+#if USE_SSL
+void TLSLog::onInvalidCertificate(const void *, Poco::Net::VerificationErrorArgs & errorCert)
+{
+    // Don't put every certificate failure into the log to save disk space and protect from misbehaving clients.
+    // Instead, produce a log per cert serial number once in `TLSLogErrorFlushIntervalSeconds` seconds.
+    std::unique_lock lock{tls_failure_stats_mutex};
+    const String &serial_number = errorCert.certificate().serialNumber();
+    if (!tls_failure_stats.contains(serial_number))
+        tls_failure_stats[serial_number] = std::make_unique<CertStats>(0, std::make_unique<Poco::Net::X509Certificate>(errorCert.certificate()));
```

Comment:
Lets define a c-tor for `CertStats`.

Reply:
When we know that sert is not calid, what information could we trust?
`errorCert.certificate().serialNumber();` -- could we trust its `serialNumber`?
You use `serialNumber` as a unique key. I doubt that it is a good choice.
Because:
- it is set by CA, if the cert is malformed, there could be anything written in
- it is not unique ID even for good certs. 
We have to use some hash from the data here as a key.

Is it save to parse its other fields for malformed certs?